### PR TITLE
media-sound/ncmpcpp: add p20240724, improve visualizer message

### PR DESCRIPTION
**Pull request by fkobi**
new patches dropped so I wanted to use them.

Also added new use flag to make curl optional. I am open to suggestions for a better name.

---

Please check all the boxes that apply:

- [X] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [X] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [X] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [X] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

### CODE REVIEW

File: media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~arm64 ~ppc ~ppc64 ~sparc ~x86"
+IUSE="clock network outputs taglib visualizer"
+
+RDEPEND="
+	>=media-libs/libmpdclient-2.1
+	dev-libs/boost:=[icu,nls]
+	sys-libs/ncurses:=[unicode(+)]
+	sys-libs/readline:=
+	network? ( net-misc/curl )
```
Comment (juippis) [inline media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild:25]: This doesn't seem to have any function in the ebuild? It's just pulling a dependency via RDEPEND which is against our policy.
https://projects.gentoo.org/qa/policy-guide/dependencies.html#pg0001

as mentioned in the policy, use an `optfeature` for deps like this. Unless you simply forgot to add the configure option to this ebuild.

Comment (NexAdn) [reply media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild:25]: I don't see any matching option when running `./configure --help` in the current master. I guess this means that this should be an optfeature.

File: media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+S="${WORKDIR}/${PN}-${NCMPCPP_COMMIT}"
+
+LICENSE="GPL-2"
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~arm64 ~ppc ~ppc64 ~sparc ~x86"
+IUSE="clock network outputs taglib visualizer"
+
+RDEPEND="
+	>=media-libs/libmpdclient-2.1
+	dev-libs/boost:=[icu,nls]
```
Comment (NexAdn) [inline media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild:22]: Why are you removing `dev-libs/icu:=` as dependency? From what I see, ncmpcpp needs it:

```
$ objdump -p /usr/bin/ncmpcpp | grep NEED
  NEEDED               libboost_filesystem.so.1.85.0
  NEEDED               libboost_locale.so.1.85.0
  NEEDED               libboost_program_options.so.1.85.0
  NEEDED               libboost_thread.so.1.85.0
  NEEDED               libicui18n.so.74
  NEEDED               libicuuc.so.74
  NEEDED               libmpdclient.so.2
  NEEDED               libreadline.so.8
  NEEDED               libncursesw.so.6
  NEEDED               libtinfow.so.6
  NEEDED               libcurl.so.4
  NEEDED               libstdc++.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libc.so.6
  VERNEED              0x000000000000b9f0
  VERNEEDNUM           0x0000000000000004
```

Also, looking at `configure.ac`, there is a call `PKG_CHECK_MODULES([ICU], [icu-i18n icu-uc], [ (...)`

Same goes for `net-libs/curl`. `configure.ac` considers a missing libcurl a fatal error, so this should probably be an unconditional dependency.

File: media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild (hunk 1 above)
Comment (NexAdn) [reply media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild:25]: Nevermind, see my above comment regarding missing dependencies.

**Review by NexAdn — CHANGES_REQUESTED** [review]

File: media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+inherit autotools
+
+NCMPCPP_COMMIT="68daf44032784a5b5b74781a9bf3826053eff4a2"
+
+DESCRIPTION="Featureful ncurses based MPD client inspired by ncmpc"
+HOMEPAGE="https://rybczak.net/ncmpcpp/"
```
Comment (NexAdn) [inline media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild:11]: Why was the GitHub upstream removed from HOMEPAGE?

File: media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild (hunk 2 above)
Comment (NexAdn) [reply media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild:22]: This is called a transitive dependency. While in this particular case you could be quite sure that this will always cause a transitive dependency to `dev-libs/icu`, in other cases this might not be true. It is good practice to always state exactly all _direct_ dependencies for a package so one can be sure that the dependencies will be pulled in.

Btw, this reasoning is similar to why paradigms like [IWYU](https://github.com/include-what-you-use/include-what-you-use?tab=readme-ov-file#instructions-for-users) exist in software development. Just the extra mental work of having to think about the implicit dependencies that are pulled in by the explicitly stated dependencies makes the ebuild harder to understand and can cause issues later on when the dependency implicitly pulling in the other dependency suddenly decides not to pull this dependency in anymore. So while this might make the ebuild shorter, I wouldn't consider it simpler and definitely not easier to maintain :sweat_smile:

File: media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild (hunk 3 above)
Comment (NexAdn) [reply media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild:11]: [Devmanual](https://devmanual.gentoo.org/ebuild-writing/variables/#ebuild-defined-variables) states that the code upstream might be listed if it's hard to find on the upstream homepage. And I haven't event found a single link to the GitHub repository on the upstream homepage. And if there is indeed a link, it is hidden pretty well.

Other than that, I personally like having the code upstream in my packages so I can easily click on the link when searching with eix, but that's just personal taste^^

File: media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild (hunk 1 above)
Comment (NexAdn) [reply media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild]: Unfortunately, no. `net-misc/curl` must be where it was before. The binary links against libcurl, so it is needed both at build time (i.e. `DEPEND`) and run time (i.e. `RDEPEND`).

Furthermore, this cannot be an optional dependency. As far as I can tell, it is linked unconditionally. Thus, the new USE flag must be removed as well.

Comment (NexAdn) [reply media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild]: You're right, the dependency is probably incorrectly labeled upstream.

File: media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+DESCRIPTION="Featureful ncurses based MPD client inspired by ncmpc"
+HOMEPAGE="
+	https://rybczak.net/ncmpcpp/
+	https://github.com/ncmpcpp/ncmpcpp/
+"
+SRC_URI="https://github.com/ncmpcpp/ncmpcpp/archive/${NCMPCPP_COMMIT}.tar.gz -> ${P}.gh.tar.gz"
+S="${WORKDIR}/${PN}-${NCMPCPP_COMMIT}"
+
+LICENSE="GPL-2"
```
Comment (mm1ke) [inline media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild:18]: can you change this to `GPL-2+`, see https://github.com/ncmpcpp/ncmpcpp/blob/master/src/settings.cpp#L6-L8

File: media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+src_configure() {
+	# --with-lto only appends -flto
+	econf \
+		$(use_enable clock) \
+		$(use_enable outputs) \
+		$(use_enable visualizer) \
+		--without-lto \
+		$(use_with taglib) \
+		$(use_with visualizer fftw)
```
Comment (mm1ke) [inline media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild:52]:
```suggestion
	local myeconfargs=(
		--without-lto
		$(use_enable clock)
		$(use_enable outputs)
		$(use_enable visualizer)
		$(use_with taglib)
		$(use_with visualizer fftw)
	)

	econf "${myeconfargs[@]}"
```

File: media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild (hunk 2 above)
Comment (juippis) [reply media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild:25]: @NexAdn explained it perfectly. We have this also written in:
https://devmanual.gentoo.org/general-concepts/dependencies/#indirect-dependencies
(page bottom)

File: media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	https://rybczak.net/ncmpcpp/
+	https://github.com/ncmpcpp/ncmpcpp/
+"
+SRC_URI="https://github.com/ncmpcpp/ncmpcpp/archive/${NCMPCPP_COMMIT}.tar.gz -> ${P}.gh.tar.gz"
+S="${WORKDIR}/${PN}-${NCMPCPP_COMMIT}"
+
+LICENSE="GPL-2+"
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~arm64 ~ppc ~ppc64 ~sparc ~x86"
+IUSE="clock -lto outputs taglib visualizer"
```
Comment (NexAdn) [inline media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild:21]: Now you have added IUSE="lto", but the corresponding option in `./configure` only triggers the `-flto` flag. Correct me if I'm wrong, but we already have `C[XX]FLAGS` for this, so this USE flag should really be necessary.

Comment (juippis) [reply media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild:21]: Correct, see 4b40de2810608796d824ccb4baed84e1a5599c79 and 783d5b43d785767b21e197a3570c857329d59f0c

Comment (juippis) [reply media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild]: No. You are literally reverting 783d5b43d785767b21e197a3570c857329d59f0c

TLDR we try to avoid using use flags to control setting `*FLAGS` as much as possible. If the `lto` has **no other function** in the ebuild, than adding `-flto` to your `*FLAGS`, then it's invalid use of use flag.

File: media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+	echo
+	elog "Example configuration files have been installed at"
+	elog "${EROOT}/usr/share/doc/${PF}"
+	elog "${P} uses ~/.ncmpcpp/config and ~/.ncmpcpp/bindings"
+	elog "as user configuration files."
+	echo
+	if use visualizer; then
+		elog "If you want to use the visualizer, mpd needs to be built with fifo USE flag."
+		echo
+	fi
```
Comment (juippis) [inline media-sound/ncmpcpp/ncmpcpp-0.9.2_p20240724.ebuild:73]: I'd use `optfeature` for this.

**Review by juippis — APPROVED** [review]
